### PR TITLE
[DNM] add push implement

### PR DESCRIPTION
**Pull request by BusyJay**
tests will be added tomorrow

### CODE REVIEW

**Comment by ngaut** [comment]
PTAL @coocood @shenli

File: src/util/xeval/evaluator.rs [hunk 1]
```diff
@@ -20,14 +20,15 @@ use std::cmp::Ordering;
...
+#[derive(Default)]
 pub struct Evaluator {
     // column_id -> column_value
     row: HashMap<i64, Datum>,
 }
 
 impl Evaluator {
-    pub fn new(row: HashMap<i64, Datum>) -> Evaluator {
-        Evaluator { row: row }
+    pub fn insert(&mut self, col_id: i64, datum: Datum) {
```
Comment (qiuyesuifeng) [inline src/util/xeval/evaluator.rs]: `Insert` is a little confused, for it is usually used in insert a row data, but here is only one column, so how about `set` or other else instead?

File: src/server/coprocessor.rs [hunk 2]
```diff
@@ -33,14 +92,299 @@ impl SnapshotCoprocessor {
...
+                    idx_info: &IndexInfo,
+                    ranges: &[KeyRange])
+                    -> Vec<KeyRange> {
+    let t_id = t_info.get_table_id();
+    let idx_id = idx_info.get_index_id();
+    if idx_id == 0 {
+        ranges.iter()
+              .map(|r| {
+                  let mut new_kr = KeyRange::new();
+                  let mut k = table::encode_row_key(t_id, r.get_start());
```
Comment (qiuyesuifeng) [inline src/server/coprocessor.rs]: Should we check the start key and end key validity?

File: src/util/xeval/evaluator.rs (hunk 1 above)
Comment (BusyJay) [inline src/util/xeval/evaluator.rs]: Evaluator only support one row, it's not a table.

File: src/server/coprocessor.rs (hunk 2 above)
Comment (BusyJay) [inline src/server/coprocessor.rs]: Coprocessor has no clue about what is a valid start key.

File: src/server/coprocessor.rs [hunk 3]
```diff
@@ -33,14 +92,299 @@ impl SnapshotCoprocessor {
...
+        }
+    } else {
+        let mut seek_key = range.take_start();
+        loop {
+            let mut res = try!(store.scan(Key::from_raw(seek_key), 1));
+            if res.is_empty() {
+                break;
+            }
+            let (key, _) = try!(res.pop().unwrap());
+            if range.get_end() < &key {
```
Comment (qiuyesuifeng) [inline src/server/coprocessor.rs]: Should we check equal here?

File: src/server/coprocessor.rs [hunk 4]
```diff
@@ -33,14 +92,299 @@ impl SnapshotCoprocessor {
...
+            return Ok(rows);
+        }
+        let h = box_try!(table::decode_handle(range.get_start()));
+        if let Some(row) = try!(get_row_by_handle(store, sel, h, eval)) {
+            rows.push(row);
+        }
+    } else {
+        let mut seek_key = range.take_start();
+        loop {
+            let mut res = try!(store.scan(Key::from_raw(seek_key), 1));
```
Comment (qiuyesuifeng) [inline src/server/coprocessor.rs]: Perhaps later we should scan batch to cache, then use filter to process it.

Comment (BusyJay) [inline src/server/coprocessor.rs]: No, actually scan is implemented by seek, and we always use a different key to seek next time.

File: src/server/coprocessor.rs [hunk 5]
```diff
@@ -33,14 +92,299 @@ impl SnapshotCoprocessor {
...
+            }
+        }
+    }
+    row.set_handle(handle);
+    if !sel.has_field_where() {
+        return Ok(Some(row));
+    }
+    if !row.get_data().is_empty() {
+        let (datums, _) = box_try!(datum::decode(row.get_data()));
+        for (c, d) in columns.iter().zip(datums) {
```
Comment (qiuyesuifeng) [inline src/server/coprocessor.rs]: Is `zip` for data compression here? What is the time cost for `zip` function call?

File: src/server/coprocessor.rs (hunk 4 above)
Comment (qiuyesuifeng) [inline src/server/coprocessor.rs]: I mean we can seek from a start key for a batch, then use next key for seek, perhaps it is an optimization, but i am not sure. Actually it is not a critical thing, only comment for mark.

**Comment by BusyJay** [comment]
PTAL

File: src/server/coprocessor.rs [hunk 6]
```diff
@@ -12,35 +12,406 @@
...
-    #[allow(dead_code)]
     fn new_snapshot<'a>(&'a self, ctx: &Context, start_ts: u64) -> Result<SnapshotStore<'a>> {
         let snapshot = try!(self.engine.snapshot(ctx));
         Ok(SnapshotStore::new(snapshot, start_ts))
     }
 }
+
+impl SnapshotEndPoint {
+    pub fn handle_select(&self, mut req: Request, sel: SelectRequest) -> Result<Response> {
+        let store = try!(self.new_snapshot(req.get_context(), sel.get_start_ts()));
```
Comment (ngaut) [inline src/server/coprocessor.rs]: s/store/snap

File: src/server/coprocessor.rs [hunk 7]
```diff
@@ -12,35 +12,406 @@
...
+        }
+        let data = box_try!(sel_resp.write_to_bytes());
+        resp.set_data(data);
+        Ok(resp)
+    }
+}
+
+fn to_pb_error(err: &Error) -> select::Error {
+    let mut e = select::Error::new();
+    e.set_code(1);
```
Comment (ngaut) [inline src/server/coprocessor.rs]: What does 1 mean ? Looks like a magic number.

**Comment by coocood** [comment]
LGTM

File: src/server/coprocessor.rs [hunk 8]
```diff
@@ -94,35 +94,32 @@ enum EndPointMessage {
     Close,
 }
 
+fn poll_messae(engine: Arc<Box<Engine>>, rx: Receiver<EndPointMessage>, ch: SendCh) {
```
Comment (ngaut) [inline src/server/coprocessor.rs]: s/poll_messae/msg_poller

**Comment by ngaut** [comment]
Rest LGTM

**Comment by ngaut** [comment]
LGTM

**Pull request merged**